From c24dd84ca64cf94cadf2da6459db214afe75b999 Mon Sep 17 00:00:00 2001
From: Peter McNeeley <petermcneeley@chromium.org>
Date: Thu, 2 Mar 2023 21:37:51 +0000
Subject: [PATCH] Reserve background to avoid reallocation of candidates

This reserve call saves about 14us.

Bug: 1396122
Change-Id: Ic643c26129f7814ed21c5754dc071e40da0625c9
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4304081
Reviewed-by: Kramer Ge <fangzhoug@chromium.org>
Commit-Queue: Peter McNeeley <petermcneeley@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1112449}
---
 .../viz/service/display/overlay_processor_delegated.cc      | 6 +++++-
 .../viz/service/display/overlay_processor_delegated.h       | 1 +
 2 files changed, 6 insertions(+), 1 deletion(-)

--- a/components/viz/service/display/overlay_processor_delegated.cc
+++ b/components/viz/service/display/overlay_processor_delegated.cc
@@ -99,6 +99,9 @@
   supports_clip_rect_ = ui::OzonePlatform::GetInstance()
                             ->GetPlatformRuntimeProperties()
                             .supports_clip_rect;
+  needs_background_image_ = ui::OzonePlatform::GetInstance()
+                                ->GetPlatformRuntimeProperties()
+                                .needs_background_image;
 }
 
 OverlayProcessorDelegated::~OverlayProcessorDelegated() = default;
@@ -158,7 +161,8 @@
       &output_color_matrix, GetPrimaryPlaneDisplayRect(primary_plane),
       is_delegated_context, supports_clip_rect_ && enable_clip_rect());
 
-  candidates->reserve(quad_list->size());
+  const auto kExtraCandiates = needs_background_image_ ? 1 : 0;
+  candidates->reserve(quad_list->size() + kExtraCandiates);
   int num_quads_skipped = 0;
 
   for (auto it = quad_list->begin(); it != quad_list->end(); ++it) {
--- a/components/viz/service/display/overlay_processor_delegated.h
+++ b/components/viz/service/display/overlay_processor_delegated.h
@@ -108,6 +108,7 @@
 
   DelegationStatus delegated_status_ = DelegationStatus::kCompositedOther;
   bool supports_clip_rect_ = false;
+  bool needs_background_image_ = false;
 };
 }  // namespace viz
 
